pkg/tasks/update_template_distributions.go

```diff
@@ -0,0 +1,200 @@
+package tasks
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	"github.com/content-services/content-sources-backend/pkg/api"
+	"github.com/content-services/content-sources-backend/pkg/config"
+	"github.com/content-services/content-sources-backend/pkg/dao"
+	"github.com/content-services/content-sources-backend/pkg/db"
+	"github.com/content-services/content-sources-backend/pkg/models"
+	"github.com/content-services/content-sources-backend/pkg/pulp_client"
+	"github.com/content-services/content-sources-backend/pkg/tasks/payloads"
+	"github.com/content-services/content-sources-backend/pkg/tasks/queue"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"golang.org/x/exp/slices"
+)
+
+func UpdateTemplateDistributionsHandler(ctx context.Context, task *models.TaskInfo, queue *queue.Queue) error {
+	opts := payloads.UpdateTemplateDistributionsPayload{}
+	if err := json.Unmarshal(task.Payload, &opts); err != nil {
+		return fmt.Errorf("payload incorrect type for UpdateTemplateDistributions")
+	}
+
+	logger := LogForTask(task.Id.String(), task.Typename, task.RequestID)
+	ctxWithLogger := logger.WithContext(ctx)
+
+	daoReg := dao.GetDaoRegistry(db.DB)
+	domainName, err := daoReg.Domain.Fetch(task.OrgId)
+	if err != nil {
+		return err
+	}
+
+	rhDomainName, err := daoReg.Domain.Fetch(config.RedHatOrg)
+	if err != nil {
+		return err
+	}
+
+	pulpClient := pulp_client.GetPulpClientWithDomain(ctxWithLogger, domainName)
+
+	t := UpdateTemplateDistributions{
+		orgId:          task.OrgId,
+		domainName:     domainName,
+		rhDomainName:   rhDomainName,
+		repositoryUUID: task.RepositoryUUID,
+		daoReg:         daoReg,
+		pulpClient:     pulpClient,
+		task:           task,
+		payload:        &opts,
+		queue:          queue,
+		ctx:            ctx,
+		logger:         logger,
+	}
+
+	return t.Run()
+}
+
+type UpdateTemplateDistributions struct {
+	orgId          string
+	domainName     string
+	rhDomainName   string
+	repositoryUUID uuid.UUID
+	daoReg         *dao.DaoRegistry
+	pulpClient     pulp_client.PulpClient
+	payload        *payloads.UpdateTemplateDistributionsPayload
+	task           *models.TaskInfo
+	queue          *queue.Queue
+	ctx            context.Context
+	logger         *zerolog.Logger
+}
+
+func (t *UpdateTemplateDistributions) Run() error {
```

Comment:
I'm not sure if they way I've architect-ed the code here is confusing to follow. The way I organized the the flow of this task is a little bit the result of me finding bugs and then fixing them. I'm very much open to other ways handling the logic here.

Reply:
One possible way i could see restructuring would be to extract 3 method (one to handle repo additions, one to handle repo updates, and one to handle repo deletions) and instead of looping over allRepos, loop over reposAdded, reposRemoved, reposUnchanged etc... in each method

```
reposAdded, reposRemoved, reposUnchanged, allRepos, err := t.daoReg.Template.GetRepoChanges(t.payload.TemplateUUID, t.payload.RepoConfigUUIDs)
```

Reply:
good idea. made that change.